src/main/java/build/buildfarm/worker/CASFileCache.java

```diff
@@ -0,0 +1,238 @@
+          // FIXME need to do this according to expiration strategy
+          Files.delete(file);
+        } else {
+          paths.put(file, new FileEntry(
+              file,
+              size,
+              attrs.lastModifiedTime().to(TimeUnit.NANOSECONDS),
+              attrs.lastAccessTime().to(TimeUnit.NANOSECONDS)));
+          sizeInBytes += size;
+        }
+        return FileVisitResult.CONTINUE;
+      }
+    });
+    update();
+  }
+
+  public void update() {
+    for (Path path : accessScanPaths) {
+      FileEntry entry = paths.get(path);
+      if (entry != null) {
+        try {
+          BasicFileAttributes attrs = Files.readAttributes(path, PosixFileAttributes.class, NOFOLLOW_LINKS);
+          /* should verify invariants, are */
+          paths.get(path).setAccessTime(attrs.lastAccessTime().to(TimeUnit.NANOSECONDS));
+        } catch (IOException ex) {
+          // FIXME report error
+        }
+      }
+    }
+    accessScanPaths.clear();
+    expirationQueue = Queues.newPriorityQueue(paths.values());
+  }
+
+  private void expireEntry() {
+    FileEntry entry = expirationQueue.remove();
+
+    Path path = entry.getPath();
+    try {
+      Files.delete(path);
+      paths.remove(path);
+      sizeInBytes -= entry.getSizeInBytes();
+    } catch (IOException ex) {
+      // FIXME report error
+    }
+  }
+
+  public Path put(Digest digest, boolean isExecutable) {
+    long size = digest.getSizeBytes();
+    Path path = root.resolve(String.format(
+        "%s_%d%s", digest.getHash(), size, isExecutable ? "_exec" : ""));
+
+    if (paths.containsKey(path)) {
+      accessScanPaths.add(path);
+      return path;
+    }
+
+    // the cache cannot hold this file
+    if (size > maxSizeInBytes) {
+      throw new IllegalStateException();
+    }
+
+    // remove cache nodes until we have cache size + digest.size_bytes <= limit
+    while (sizeInBytes + size > maxSizeInBytes && !expirationQueue.isEmpty()) {
+      // FIXME could have no entries left to remove...
+      expireEntry();
+    }
+
+    if (sizeInBytes + size > maxSizeInBytes) {
+      System.out.println(String.format(
+          "Expiration did not yield sufficient space: %d + %d > %d",
+          sizeInBytes,
+          size,
+          maxSizeInBytes));
+      return null;
+    }
+
+    try (InputStream in = instance.newStreamInput(instance.getBlobName(digest))) {
+      // FIXME make a validating file copy object and verify digest
+      Path tmpPath = path.resolveSibling(path.getFileName() + ".tmp");
+      long copySize = Files.copy(in, tmpPath);
+      in.close();
+      if (copySize != size) {
+        Files.delete(tmpPath);
+        System.out.println(String.format(
+            "Invalid file size copied into cache: was %d, expected %d",
+            copySize,
+            size));
+        return null;
+      }
+      setPermissions(tmpPath, isExecutable);
+      Files.move(tmpPath, path, REPLACE_EXISTING);
```

Comment:
Do we need to fsync the file before moving?

Reply:
Not for any posix filesystem I've ever worked with for close->rename sequencing? The sibling construction is purposeful and the resulting syscall generation seems to indicate java's compliance.